flocker/acceptance/testtools.py

```diff
@@ -703,6 +729,59 @@ def current_nodes(self):
         return request
 
     @log_method
+    def install_flocker_version(self, package_source):
+        """
+        Change the version of flocker installed on all of the nodes to the
+        version indicated by `package_source`.
+        """
+        control_node_address = self.control_node.public_address
+        all_cluster_nodes = set([x.public_address for x in self.nodes] +
+                                [control_node_address])
+        distribution = self.distribution
+
+        def get_flocker_version():
+            # Retry getting the flocker version for 10 seconds. Flocker might
+            # not be running yet on the control node when this is called.
+            d = timeout(
+                self.reactor,
+                retry_failure(
+                    self.reactor,
+                    self.client.version
+                ),
+                10
+            )
```

Comment:
If the connection hangs indefinitely that will hang indefinitely. If we have a chance to break an indefinitely hanging connection, deferred cancellation is our best chance.

I've seen a few of the benchmarking folks start to use this pattern: for instance, the in-flight: https://github.com/ClusterHQ/flocker/pull/2581/files#diff-5c624e764d7d12f04e516a85cae8ff77R39

Room for disagreement about timing out each retry over timing out the whole set of retries. Not sure I have strong opinions either way. This will for sure finish in 10 seconds, that would for sure finish in timeout*steps seconds. Some data collection about what the failure modes are like would probably be required to make an educated decision.